File: frontend/src/api/webhooks.ts
/**
 * API Webhooks - Gestion des webhooks pour les événements temps réel
 */

import api from '../services/api'

export interface Webhook {
  id: string
  url: string
  events: string[]
  description?: string
  is_active: boolean
  last_triggered_at?: string
  consecutive_failures: number
  created_at: string
}

export interface CreateWebhookRequest {
  url: string
  events: string[]
  description?: string
}

export interface CreateWebhookResponse {
  webhook: Webhook
  secret: string // Affiché UNE SEULE FOIS
}

export interface WebhookDelivery {
  id: string
  webhook_id: string
  event_type: string
  attempt: number
  success: boolean
  status_code?: number
  response_time_ms?: number
  error_message?: string
  delivered_at: string
}

/**
 * Service pour gérer les webhooks.
 */
export const webhooksApi = {
  /**
   * Liste tous les webhooks de l'utilisateur.
   *
   * @returns Liste des webhooks
   */
  async list(): Promise<Webhook[]> {
    const response = await api.get<{ total: number; webhooks: Webhook[] }>('/api/v1/webhooks')
    return response.data.webhooks
  },

  /**
   * Crée un nouveau webhook.
   *
   * IMPORTANT: Le secret est retourné UNE SEULE FOIS.
   * Il doit être copié immédiatement car il ne sera plus accessible.
   *
   * @param data - Données de création
   * @returns Webhook créé avec secret (une fois)
   */
  async create(data: CreateWebhookRequest): Promise<CreateWebhookResponse> {
    const response = await api.post<{
      id: string
      url: string
      events: string[]
      description?: string
      secret: string
      created_at: string
    }>('/api/v1/webhooks', data)

    // Transformer la réponse backend pour correspondre à l'interface frontend
    return {
      secret: response.data.secret,
      webhook: {
        id: response.data.id,
        url: response.data.url,
        events: response.data.events,
        description: response.data.description,
        is_active: true,
        last_triggered_at: undefined,
        consecutive_failures: 0,
        created_at: response.data.created_at,
      },
    }
  },

  /**
   * Supprime un webhook.
   *
   * @param id - UUID du webhook à supprimer
   */
  async delete(id: string): Promise<void> {
    await api.delete(`/api/v1/webhooks/${id}`)
  },

  /**
   * Teste un webhook en envoyant un événement de test.
   *
   * @param id - UUID du webhook à tester
   */
  async test(id: string): Promise<void> {
    await api.post(`/api/v1/webhooks/${id}/test`)
  },

  /**
   * Récupère l'historique des deliveries d'un webhook.
   *
   * @param id - UUID du webhook
   * @param limit - Nombre de deliveries à récupérer
   * @returns Liste des deliveries
   */
  async deliveries(id: string, limit = 50): Promise<WebhookDelivery[]> {
    const response = await api.get<{ total: number; deliveries: WebhookDelivery[] }>(
      `/api/v1/webhooks/${id}/deliveries?limit=${limit}`
    )
    return response.data.deliveries
  },
}

export default webhooksApi

File: sdk/python/hub_chantier/resources/webhooks.py
"""Ressource Webhooks."""

from typing import List, Dict, Optional, Any
from .base import BaseResource


class Webhooks(BaseResource):
    """Gestion des webhooks."""

    def list(self) -> List[Dict[str, Any]]:
        """
        Liste tous les webhooks.

        Returns:
            Liste de webhooks
        """
        return self.client._request("GET", "/api/v1/webhooks")  # type: ignore

    def create(
        self, url: str, events: List[str], description: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Crée un nouveau webhook.

        Args:
            url: URL de callback (doit être HTTPS)
            events: Liste d'événements à écouter (ex: ["chantier.created"])
            description: Description optionnelle

        Returns:
            Webhook créé avec secret

        Example:
            >>> webhook = client.webhooks.create(
            ...     url="https://myapp.com/webhooks/hub-chantier",
            ...     events=["chantier.created", "affectation.created"],
            ...     description="Production webhook"
            ... )
            >>> secret = webhook['secret']  # À sauvegarder pour vérifier signatures
        """
        data: Dict[str, Any] = {"url": url, "events": events}
        if description:
            data["description"] = description

        return self.client._request("POST", "/api/v1/webhooks", json=data)

    def delete(self, webhook_id: str) -> None:
        """
        Supprime un webhook.

        Args:
            webhook_id: ID du webhook (UUID)
        """
        self.client._request("DELETE", f"/api/v1/webhooks/{webhook_id}")
